src/language/language_type.rs

```diff
@@ -19,23 +19,31 @@ impl LanguageType {
     /// Parses a given `Path` using the `LanguageType`. Returning `Stats`
     /// on success.
     pub fn parse(self, path: PathBuf) -> io::Result<Stats> {
+        self.parse_with_accuracy(path).map_err(|e| e.0)
+    }
+
+    /// Parses a given `Path` using the `LanguageType`. Returning `Stats`
+    /// on success and giving back ownership of PathBuf on error.
+    pub fn parse_with_accuracy(self, path: PathBuf) -> Result<Stats, (io::Error, PathBuf)> {
         let text = {
-            let f = File::open(&path)?;
+            let f = match File::open(&path) {
+                Ok(f) => f,
+                Err(e) => return Err((e, path)),
+            };
             let mut s = String::new();
             let mut reader = DecodeReaderBytes::new(f);
 
-            reader.read_to_string(&mut s)?;
+            if let Err(e) = reader.read_to_string(&mut s) {
+                return Err((e, path));
+            }
             s
         };
-
-        self.parse_from_str(path, &text)
+        Ok(self.parse_from_str(path, &text).unwrap())
```

Comment:
Oh right, I did this cause the results have different error cases and parse_from_str cannot fail from its implementation. Probably a bad idea to unwrap though since the implementation could change. Will have to do the same if let Err instead of map_err though due to the closure move that would happen otherwise though

Reply:
Actually i have a problem here in general cause the parse_from_str error that wouldnt happen in its current implementation doesnt give back the pathbuf, so I am not sure how to even map the error here